### PR TITLE
BigQuery connector

### PR DESCRIPTION
**Pull request by Rafalz13**
## Summary
The connector which allows connecting to BigQuery data ( to project, datasets, tables).

## Importance
It is the new source from which we can extract the data. 

## Checklist

This PR:

- [x] adds new tests (if appropriate)
- [x] updates docstrings for any new functions or function arguments (if appropriate)
- [x] updates `CHANGELOG.md` with a summary of the changes

### CODE REVIEW

File: tests/unit/test_bigquery.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import pytest
```
Comment (angelika233) [inline tests/unit/test_bigquery.py:1]: remove unused import

File: tests/unit/test_bigquery.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+from viadot.sources import BigQuery
+
+
+BIGQ = BigQuery(credentials_key="BIGQUERY_TESTS")
```
Comment (angelika233) [inline tests/unit/test_bigquery.py:5]: Maybe file should be moved to integration because it uses local credentials

File: viadot/flows/bigquery_to_adls.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    ):
+        """
+        Flow for downloading data from BigQuery project to a local CSV or Parquet file
+        using Bigquery API, then uploading it to Azure Data Lake.
+
+        Args:
+            name (str, optional): _description_. Defaults to None.
+            project (str, optional): Project name. Defaults to None.
+            dataset (str, optional): Dataset name. Defaults to None.
+            table (str, optional): Table name . Defaults to None.
```
Comment (angelika233) [inline viadot/flows/bigquery_to_adls.py:59]: remove space before dot

File: viadot/flows/bigquery_to_adls.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+            credentials_key (str, optional): Credential key to dictionary where details are stored. Defaults to "BIGQUERY".
+            credentials (dict, optional): Credentials dictionary - credentials can be generate as key
```
Comment (angelika233) [inline viadot/flows/bigquery_to_adls.py:61]: maybe "generated" instead "generate"

File: viadot/flows/bigquery_to_adls.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+            end_date (str, optional): A query parameter to pass end date e.g. "2022-01-01". Defaults to None.
+            output_file_extension (str, optional): Output file extension - to allow selection of .csv for data
+            which is not easy to handle with parquet. Defaults to ".parquet".
+            adls_dir_path (str, optional): Azure Data Lake destination folder/catalog path. Defaults to None.
+            local_file_path (str, optional): Local destination path. Defaults to None.
+            adls_file_name (str, optional): Name of file in ADLS. Defaults to None.
+            adls_sp_credentials_secret (str, optional): The name of the Azure Key Vault secret containing a dictionary with
+            ACCOUNT_NAME and Service Principal credentials (TENANT_ID, CLIENT_ID, CLIENT_SECRET) for the Azure Data Lake.
+            Defaults to None.
+            if_exists (str, optional): Whether the file exists. Defaults to "replace".
```
Comment (angelika233) [inline viadot/flows/bigquery_to_adls.py:73]: Maybe "What to do if the file exists"  would be better than "Whether"

File: viadot/tasks/bigquery.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        **kwargs,
+    ):
+        """
+        Initialize BigQueryToDF object.
+        For querying on database dataset name and table name is needed.
+
+        Args:
+            project (str, optional): Project name. Defaults to None.
+            dataset (str, optional): Dataset name. Defaults to None.
+            table (str, optional): Table name . Defaults to None.
```
Comment (angelika233) [inline viadot/tasks/bigquery.py:36]: remove space

File: viadot/tasks/prefect.py [hunk 7]
```diff
@@ -201,12 +201,16 @@ def run(
         """
             % flow_name
         )
-
+        print(f"Before -> '{flow_name}' date_range_type = {date_range_type}")
```
Comment (angelika233) [inline viadot/tasks/prefect.py:204]: remove print

File: viadot/tasks/prefect.py [hunk 8]
```diff
@@ -201,12 +201,16 @@ def run(
...
         client = prefect.Client()
         flow_runs = client.graphql(query)
         flow_runs_details = flow_runs.data.flow
-
         time_schedule = flow_runs_details[0]["flow_runs"][0]["scheduled_start_time"]
+        # for flow_run in iter_throught_flow_runs(flow_runs_details=flow_runs_details):
```
Comment (angelika233) [inline viadot/tasks/prefect.py:209]: Are these comments needed?

File: viadot/tasks/prefect.py [hunk 9]
```diff
@@ -222,7 +226,9 @@ def run(
             new_date_range_type = self.change_date_range(
                 date_range=date_range_type, difference=difference_days
             )
+            print(f"After -> '{flow_name}' date_range_type = {new_date_range_type}")
```
Comment (angelika233) [inline viadot/tasks/prefect.py:229]: remove print

File: viadot/tasks/prefect.py [hunk 10]
```diff
@@ -222,7 +226,9 @@ def run(
...
             return new_date_range_type
 
         if is_scheduled is False:
+            print("FLOW was not scheduled")
```
Comment (angelika233) [inline viadot/tasks/prefect.py:233]: remove print

File: tests/unit/test_bigquery.py [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    query_job = BIGQ.query(query)
+    for row in query_job:
+        names[row["name"]] = row["total"]
+
+    assert names == {
+        "James": 4924235,
+        "John": 4818746,
+        "Robert": 4703680,
+        "Michael": 4280040,
+    }
```
Comment (angelika233) [inline tests/unit/test_bigquery.py:42]: add tests for bigquery_to_adls if possible

File: viadot/tasks/bigquery.py [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+        For querying on database - dataset and table name is required.
+
+        Args:
+            project (str, optional): Project name - taken from the json file (project_id). Defaults to None.
+            dataset (str, optional): Dataset name. Defaults to None.
+            table (str, optional): Table name. Defaults to None.
+            date_column_name (str, optional): The query is based on a date, the user can provide the name
+            of the date columnn if it is different than "date". If the user-specified column does not exist,
+            all data will be retrieved from the table. Defaults to "date".
+            for User Principal inside a BigQuery project. Defaults to None.
```
Comment (m-paz) [inline viadot/tasks/bigquery.py:41]: seems that arg is missing

File: viadot/tasks/bigquery.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+        project: str = None,
+        dataset: str = None,
+        table: str = None,
+        date_column_name: str = "date",
+        start_date: str = None,
+        end_date: str = None,
+        credentials_key: str = "BIGQUERY",
+        **kwargs,
+    ) -> None:
+        bigq = BigQuery(credentials_key=credentials_key)
```
Comment (m-paz) [inline viadot/tasks/bigquery.py:85]: maybe it's a better idea to keep the full name like bigquery?

File: viadot/tasks/bigquery.py [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+            and table in bigq.list_tables(dataset)
+        ):
+            try:
+                if start_date is not None and end_date is not None:
+                    query = f"""SELECT * FROM `{dataset}.{table}` 
+                    where {date_column_name} between PARSE_DATE("%Y-%m-%d", "{start_date}") and PARSE_DATE("%Y-%m-%d", "{end_date}") 
+                    order by {date_column_name} desc"""
+                else:
+                    query = f"""SELECT * FROM `{project}.{dataset}.{table}` 
+                    where {date_column_name} < CURRENT_DATE()
```
Comment (m-paz) [inline viadot/tasks/bigquery.py:103]: always today -1? It's fine but please put this info in the doc string.

File: viadot/tasks/prefect.py [hunk 15]
```diff
@@ -36,7 +36,7 @@ def get_time_from_last_successful_run(flow_runs_details: List[dict] = None) -> s
     """
 
     for flow_run in iter_throught_flow_runs(flow_runs_details=flow_runs_details):
-        if flow_run["state"] == "Success":
+        if flow_run["state"] == "Failed":
```
Comment (m-paz) [inline viadot/tasks/prefect.py:39]: what is the reason of this change?

**Review by m-paz — COMMENTED** [review]
please check the comments within the code

File: viadot/sources/bigquery.py [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+        Create an instance of BigQuery class.
+
+        Args:
+            credentials_key (str): Credential key to dictionary where details are stored. Credentials can be generated as key
+            for User Principal inside a BigQuery project. Defaults to None.
+
+        Raises:
+            CredentialError: In case credentials cannot be found.
+        """
+        self.path_json = "credentials_json.json"
```
[the remark this answers is not in the document]
Comment (Rafalz13) [reply viadot/sources/bigquery.py:31]: I took it from the local config but I extracted only one key 'BIGQUERY' - not all credentials which are inside. Then I convert it to a json file. This type of authentication needs a json file with credentials to big query.

File: viadot/sources/bigquery.py [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+        credentials = local_config.get(credentials_key)
+        if credentials is None:
+            raise CredentialError("Credentials not found.")
+
+        super().__init__(*args, credentials=credentials, **kwargs)
+
+        with open(self.path_json, "w") as json_file:
```
[the remark this answers is not in the document]
Comment (Rafalz13) [reply viadot/sources/bigquery.py:38]: Commented above

**Review by m-paz — APPROVED** [review]

File: viadot/sources/bigquery.py [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+        Raises:
+            CredentialError: In case credentials cannot be found.
+        """
+        self.credentials_raw = local_config.get(credentials_key)
+        if self.credentials_raw is None:
+            raise CredentialError("Credentials not found.")
+
+        super().__init__(*args, credentials=self.credentials_raw, **kwargs)
+
+        credentials = service_account.Credentials.from_service_account_info(
```
Comment (m-paz) [inline viadot/sources/bigquery.py:36]: We should add the option for Azure KV storage - it's not mandatory but it should be added in the next version or even now.

File: viadot/tasks/bigquery.py [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+        *args,
+        **kwargs,
+    ):
+        """
+        Initialize BigQueryToDF object. For querying on database - dataset and table name is required.
+        The project name is taken from config/credential json file.
+
+        There are 3 cases:
+            If start_date and end_date are not None - all data from the start date to the end date will be retrieved.
+            If start_date and end_date are left as default (none) - the data is pulled to "yesterday" (current date -1)
```
Comment (m-paz) [inline viadot/tasks/bigquery.py:35]: it's just grammar - "till yesterday"

File: viadot/flows/bigquery_to_adls.py [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+        output_file_extension: str = ".parquet",
+        adls_dir_path: str = None,
+        local_file_path: str = None,
+        adls_file_name: str = None,
+        adls_sp_credentials_secret: str = None,
+        if_exists: str = "replace",
+        *args: List[any],
+        **kwargs: Dict[str, Any],
+    ):
+        """
```
Comment (m-paz) [inline viadot/flows/bigquery_to_adls.py:50]: maybe it will be worth to put also here the info about date ranges ( these 3 scenarios yesterday etc)

**Review by m-paz — COMMENTED** [review]
+ please resolve the conflict
